### PR TITLE
SONARJS-578 Improve type-inference: consider string methods

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    putString("charAt", PrimitiveType.STRING);
+    putString("charCodeAt", PrimitiveType.NUMBER);
+    putString("concat", PrimitiveType.STRING);
+    putString("indexOf", PrimitiveType.NUMBER);
+    putString("lastIndexOf", PrimitiveType.NUMBER);
+    putString("localeCompare", PrimitiveType.NUMBER);
+    putString("replace", PrimitiveType.STRING);
+    putString("search", PrimitiveType.NUMBER);
+    putString("slice", PrimitiveType.STRING);
+    // fixme. Should be created unique object for each "split" call
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import org.sonar.plugins.javascript.api.tree.expression.CallExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.expression.MemberExpressionTree;
+
+public class BuiltInMethods {
+
+  private static final Map<String, Type> STRING_METHOD_RETURN_TYPES = new HashMap<>();
+
+  static {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java]: @vilchik-elena Another way to initialize such a Map is: `ImmutableMap.builder().put(key, value)....build()`

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java]: @pynicolas Thanks! fixed in 3575ba52109c37670a5097aa9e9e5b5cb2d56bb8

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    .put("charAt", PrimitiveType.STRING)
+    .put("charCodeAt", PrimitiveType.NUMBER)
+    .put("concat", PrimitiveType.STRING)
+    .put("indexOf", PrimitiveType.NUMBER)
+    .put("lastIndexOf", PrimitiveType.NUMBER)
+    .put("localeCompare", PrimitiveType.NUMBER)
+    .put("replace", PrimitiveType.STRING)
+    .put("search", PrimitiveType.NUMBER)
+    .put("slice", PrimitiveType.STRING)
+    // fixme. Should be created unique object for each "split" call
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    .put("charAt", PrimitiveType.STRING)
+    .put("charCodeAt", PrimitiveType.NUMBER)
+    .put("concat", PrimitiveType.STRING)
+    .put("indexOf", PrimitiveType.NUMBER)
+    .put("lastIndexOf", PrimitiveType.NUMBER)
+    .put("localeCompare", PrimitiveType.NUMBER)
+    .put("replace", PrimitiveType.STRING)
+    .put("search", PrimitiveType.NUMBER)
+    .put("slice", PrimitiveType.STRING)
+    // fixme. Unique object should be created for each "split" call
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/BuiltInMethods.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**